### PR TITLE
Homework

### PR DESCRIPTION
**Pull request by 0le8**

### CODE REVIEW

File: ftracker.go [hunk 1]
```diff
@@ -45,22 +45,22 @@ func meanSpeed(action int, duration float64) float64 {
 // trainingType string — вид тренировки(Бег, Ходьба, Плавание).
 // duration float64 — длительность тренировки в часах.
 func ShowTrainingInfo(action int, trainingType string, duration, weight, height float64, lengthPool, countPool int) string {
-    // ваш код здесь
-    switch {
+	// ваш код здесь
```
Comment (dirtymew) [inline ftracker.go]: Не ошибка, но такие комментарии из задания лучше убирать

File: ftracker.go [hunk 2]
```diff
@@ -45,22 +45,22 @@ func meanSpeed(action int, duration float64) float64 {
...
+	switch {
 	case trainingType == "Бег":
-		distance := ... // вызовите здесь необходимую функцию
-		speed := ... // вызовите здесь необходимую функцию
-		calories := ... // вызовите здесь необходимую функцию
+		distance := distance(action)                               // вызовите здесь необходимую функцию
```
Comment (dirtymew) [inline ftracker.go]: тут можно лучше - вынести общий код выше `switch` - это позволит избежать лишнего дублирования.

существуют принципы и практики разработки ПО направленные в конечном итоге на уменьшение кол-ва ошибок. Дублирование кода не всегда плохо - но в основном стоит его избегать.

вводная статья про это https://habr.com/ru/companies/itelma/articles/546372/

File: ftracker.go [hunk 3]
```diff
@@ -81,14 +81,16 @@ const (
 // weight float64 — вес пользователя.
 // duration float64 — длительность тренировки в часах.
 func RunningSpentCalories(action int, weight, duration float64) float64 {
-    // ваш код здесь
-    ...
+	// ваш код здесь
+	// Formula - ((18 * СредняяСкоростьВКм/ч * 1.79) * ВесСпортсменаВКг / mInKM * ВремяТренировкиВЧасах * minInH)
```
Comment (dirtymew) [inline ftracker.go]: данный комментарий лучше вынести в описание функции.
и не понятно что такое Formula - это переменная или функция ?

File: ftracker.go [hunk 4]
```diff
@@ -81,14 +81,16 @@ const (
...
+	avgSpeed := meanSpeed(action, duration)
```
Comment (dirtymew) [inline ftracker.go:84]: Хорошо -  вынес часть формулы в отдельную переменную

File: ftracker.go [hunk 5]
```diff
@@ -81,14 +81,16 @@ const (
...
+	return (runningCaloriesMeanSpeedMultiplier * avgSpeed * runningCaloriesMeanSpeedShift) * weight / mInKm * duration * minInH
```
Comment (dirtymew) [inline ftracker.go:85]: тут возможно деление на 0 - нужно исправить и избежать подобной ситуации( дальше тоже)

File: ftracker.go [hunk 6]
```diff
@@ -100,14 +102,18 @@ const (
 // weight float64 — вес пользователя.
 // height float64 — рост пользователя.
 func WalkingSpentCalories(action int, duration, weight, height float64) float64 {
-    // ваш код здесь
-    ...
+	// ваш код здесь
+	// Formula - ((0.035 * ВесСпортсменаВКг + (СредняяСкоростьВМетрахВСекунду**2 / РостВМетрах)* 0.029 * ВесСпортсменаВКг) * ВремяТренировкиВЧасах * minInH)
+	avgSpeedInMetrSec := meanSpeed(action, duration) * kmhInMsec
+	fmt.Println(avgSpeedInMetrSec)
```
Comment (dirtymew) [inline ftracker.go]: зачем тут fmt.Println ? забыл убрать...

чтобы не дебажить принтами лучше пользоваться дебагером и unit тестами

**Review by dirtymew — CHANGES_REQUESTED** [review]
Привет Олег.
Работа хорошая,  все тесты проходят - но есть что исправить - пока принять немогу.

Нужно исправить:
 - избежать деления на 0
 - убрать лишний Println
 - убрать ненужны комментарии

Что можно улучшить:
 - избежать дублирования кода
 - написать тесты на краевый случаи(например когда все параметры 0)

File: ftracker.go (hunk 6 above)
Comment (0le8) [reply ftracker.go]: Ок. Использовал для дебага, забыл удалить. Удалил в коммите [fix comment; fix division by zero; add test null](https://github.com/0le8/go-sprint-four/pull/3/commits/3c6f57e77609b0e595f36fe247db8eb3c6ed6741)

File: ftracker.go (hunk 5 above)
Comment (0le8) [reply ftracker.go:85]: Добавил проверку в коммите [fix comment; fix division by zero; add test null](https://github.com/0le8/go-sprint-four/pull/3/commits/3c6f57e77609b0e595f36fe247db8eb3c6ed6741)

File: ftracker.go (hunk 3 above)
Comment (0le8) [reply ftracker.go]: Удалил в коммите [fix comment; fix division by zero; add test null](https://github.com/0le8/go-sprint-four/pull/3/commits/3c6f57e77609b0e595f36fe247db8eb3c6ed6741)

File: ftracker.go (hunk 2 above)
Comment (0le8) [reply ftracker.go]: Вынес в функцию в коммите [fix comment; fix division by zero; add test null](https://github.com/0le8/go-sprint-four/pull/3/commits/3c6f57e77609b0e595f36fe247db8eb3c6ed6741)

File: ftracker.go (hunk 1 above)
Comment (0le8) [reply ftracker.go]: Комментарии убрал в коммите [fix comment; fix division by zero; add test null](https://github.com/0le8/go-sprint-four/pull/3/commits/3c6f57e77609b0e595f36fe247db8eb3c6ed6741)

**Comment by 0le8** [comment]
Постарался устранить все недочёты в коммите  [fix comment; fix division by zero; add test null](https://github.com/0le8/go-sprint-four/pull/3/commits/3c6f57e77609b0e595f36fe247db8eb3c6ed6741)

File: ftracker_test.go [hunk 7]
```diff
@@ -162,6 +168,48 @@ func TestShowTrainingInfo(t *testing.T) {
 			},
 			want: unknownTest,
 		},
+
+		{
+			name: "run test null",
```
Comment (dirtymew) [inline ftracker_test.go:173]: Супер - теперь не будет никаких NaN и Inf )

File: ftracker.go [hunk 8]
```diff
@@ -100,14 +100,18 @@ const (
 // weight float64 — вес пользователя.
 // height float64 — рост пользователя.
 func WalkingSpentCalories(action int, duration, weight, height float64) float64 {
-    // ваш код здесь
-    ...
+	avgSpeedInMetrSec := meanSpeed(action, duration) * kmhInMsec
+	heightInCm := height / cmInM
+	if heightInCm == 0 {
```
Comment (dirtymew) [inline ftracker.go:105]: Отлично - добавил проверку на 0,
но можно лучше: если height == 0 то мы сразу возвращаем 0 - нам не нужно выполнять деление...

File: ftracker.go [hunk 9]
```diff
@@ -45,22 +45,19 @@ func meanSpeed(action int, duration float64) float64 {
 // trainingType string — вид тренировки(Бег, Ходьба, Плавание).
 // duration float64 — длительность тренировки в часах.
 func ShowTrainingInfo(action int, trainingType string, duration, weight, height float64, lengthPool, countPool int) string {
-    // ваш код здесь
-    switch {
+	distance := distance(action)
```
Comment (dirtymew) [inline ftracker.go:48]: Отлично убрал часть дублирущего кода.
Но можно лучше speed тоже вычислить сразу а в кейсе Плавание его переопределить.

**Review by dirtymew — APPROVED** [review]
Олег

Все моменты исправлены - тесты зеленые. Отлично - зачет.

Молодец что не поленился и добавил тесты.
Есть несколько доп комментариев - а так все супер. Удачи в обучении.
